server/src/auth/token_validator.rs

```diff
@@ -75,23 +82,27 @@ impl TokenValidator {
                     }
                 })
                 .collect();
-            self.edge_sink.sink_tokens(tokens_to_sink.clone()).await?;
-            Ok(tokens_to_sink.merge(known_tokens))
+            tokens_to_sink.iter().for_each(|t| {
+                self.token_cache.insert(t.token.clone(), t.clone());
+            });
+            let updated_tokens = tokens_to_sink.upsert(known_tokens);
+            if let Some(persist) = self.persistence.clone() {
+                let _ = persist.save_tokens(updated_tokens.clone()).await;
```

Comment:
Not completely sure what you mean here? Are you saying that we shouldn't save here or that this should be processed by a different thread, lazily in the background?

Reply:
Kind of both 🤔 
1. Do we really need to save to the persistent store here?
2. If we do, can it be lazy instead?

I think only handling a read from the persistent store at load and a write at shutdown would be ideal, but maybe I'm missing something you guys already thought of.

Reply:
I want to say that it's worth it to have incremental persistence of the state, if we only include saving on the shutdown hook then we'll never persist the data in case of non graceful shutdown, which can happen for a whole bunch of valid reasons.

The writes here aren't on the hot path so they should only cause a lock on updating new tokens/features, which is a rare operation.

You may be seeing something we're not (that's why we called you!), what's worrying you about this?

Reply:
That's fair, if you feel like this is the way to go I trust you here.

I guess I'm concerned about:

- Separation of concerns / responsibility - Our inner logic should handle the state / cache layer only, without having to worry about persistence. E.g. when developing a new feature, how do I know whether I should persist the data? Should I even be worried about it at all? Why does my logic need to await these writes? Why do we need to pass it in as app data?
- Performance - Even if it doesn't happen too often, we may still be unnecessarily writing to our store, which can be nice to avoid if we can;

Are we worried about ungraceful shutdowns? How destructive would they really be? 

Another option could be to have an async task syncing to store every X minutes for example - Achieves separation of concerns while the threads operating our inner logic don't need to await for the writes to be done. But from my perspective we may not even need this at this point, as ungraceful shutdowns don't feel that destructive to me.

Probably I'm just overthinking at this point. It's not something that should block this PR anyways.

Reply:
> Are we worried about ungraceful shutdowns? How destructive would they really be?

I'm always worried about ungraceful shutdowns, they're part of life and our software shouldn't be worse off if bad things happen, not if we can help it at least.

I'm not sure the performance here is a worry but I can see the separation of concerns being a valid issue here. In practice, you shouldn't have to worry about this, outside of working on anything that touches the way features or tokens are resolved from Unleash. I'm open to the lazy caching but probably in another PR, alternatively, we could just abstract away the caching into a separate set of logic. I don't have strong feelings about when we persist, only that it's done during the lifetime of the application and regularly

Reply:
So, we've moved to persisting lazily, so there's now a proper SoC here. Great suggestion Nuno. It also allows us to not have to worry about two requests with different tokens coming in and clobbering the files.